### ISSUE TITLE
GAN128 gets worse and develops moiré patter

### ISSUE THREAD

**Issue by andenixa** [deepfakes/faceswap#339]
When training with GAN128 after a certain number of epochs (about 15hours of training on modern GPU)

- result becomes worse: faces more disfigured than at the first stages of training (both discriminator and generator).
- as GAN very quickly learns to swap eyes and mouth extremely accurately it  gradually losses the ability as more epochs pass instead of improving further.
- develops horrible moire pattern at both  mask and the output (even raw) non existent at the data-sets
(in fact data starts to look 64x64 printed at 128x128 tiles)

That happens irrespective of training patterns which are taken from various source for both dataA and dataB. I tried tv recoding, movie flics, 3d cartoons, my own footage of me and friends. The behaviors is always the same.

**Comment by rifterbater** [comment]
Missed a spot in my ugly hacking. Need to also delete this line to drop the mask during conversion:

new_face = mask * new_face + (1 - mask) * normalized_face

**Comment by andenixa** [comment]
Thanks a lot a I shall report with my findings

**Comment by andykdy** [comment]
When the size of the kernel and stride don't play well with each other it can lead to checkerboard-like convolution artifacts in the resulting image. [This article](https://distill.pub/2016/deconv-checkerboard/) can give some insight into why this happens. 
I am thinking that perhaps the line of code below is why we are seeing the moire patterns. 
``x = SeparableConv2D(f, kernel_size=3, strides=2, kernel_initializer=conv_init, use_bias=False, padding="same")(x)``
Line 66 of plugins/Model_GAN128/Model.py

**Comment by andenixa** [comment]
Thanks everyone for you feedback. 
Guys I realized while I was "tweaking" the Original model that GAN encoder doesn't have enough memory/neurons to hold the data (encoded_dim). GAN is also one Conv2d layer big to be trainable. My experiments shown that these kind of architecture can converge relatively well with 5 Conv2D layers which is 1 more than Original and one less than GAN. Also my unproven speculation that initial Conv2D layers should have a bigger windows which should span horizontal (something like (5,6) they are inverted). I haven't tested this yet, but I can see results are too generalized on horizontal features, but that yet to be proven.

PS: My current Original model tweak is times faster than GAN and gives a better result. It converges even with different versions of face such as 3D, photos, sketches, and drawings mixed together as target_B. My Original model tweak also plays nice with one-to-many approach. Only problem I found it gives too much preference to colors in favor of the shape (ie it would try to match, say, yellowish targets with photos taken in sunlight). If someone knows how to modify this part that would be very welcome. In general its OK, but the emphasys is a little to big.

**Comment by luiscosio** [comment]
@andenixa Are you planning on doing a pull request of your OriginalHighRes model?

**Comment by torzdf** [comment]
He already has. You can test it by checking out the staging branch https://github.com/deepfakes/faceswap/tree/staging

**Comment by kvrooman** [comment]
Going to post a model update in a day or so with an improvement that should alleviate or solve this issue by implementing ICNR initialization of the convolution kernels.

See this paper by the same authors of Pixel Shuffle for an overview
https://arxiv.org/ftp/arxiv/papers/1707/1707.02937.pdf

**Comment by luiscosio** [comment]
Been playing with your model and works pretty well. I'll test your update
as soon as you release it.

**Comment by tjess78** [comment]
@kvrooman any news? I'm burning for a test 'cause GAN works best with 2 of my facesets until moiré kicks in

**Comment by andenixa** [comment]
I noticed lr=1e-04 sometimes contributes to long lasting moire patterns (beats me why) (instead of 5e-05 in faceswap) and since we have additional layer of mask the cross-talk and sometimes becomes permanent.
I've can observe the same pattern with Adam lr=1e-4 at my mask-less Model though it tends to vanish with time.

**Comment by Kirin-kun** [comment]
I tried OriginalHighRes and it takes a lot longer to train, but it's expected.

But it seems to develop a moire pattern too. I'm around 0.03x loss though, so I'll see if it improves when loss goes down.

**Comment by andenixa** [comment]
@Kirin-kun I am going to make it slightly faster. I wouldn't watch for loss as its not really an objective measurement of completion. As long as you see your final result has eyes and pupil and facial expression follows the original during convert its pretty much done. As for moire I shall try to get rid of it.

Once I trained a net at 50k epochs I had all blurry silhouette of teeth and eyes with 0.04 loss, at 200k epochs I had crisp eyes and every tooth visible with clear tongue and.. well loss was still 0.04

**Comment by andenixa** [comment]
@tjess78 in my experience every 128x model is able to generate realistic consistent eyeballs.
To achieve that all that is needed is to raise Dense layer by about 50% and add additional conv2d layer prior to the last one. Train it at -bs 16 or more for ~120k epochs and you shall get consistent realistic eyeballs with every possible model.

**Comment by andenixa** [comment]
@agilebean its regular seamless mode on the left and masked with slight smoothing at the right side.
Both modes use histogram matching. I use a slight mask erosion (about 10-15%). That's pretty much everything. 
I suggest to not use any sharpening parameter to avoid the mask to be visible.

**Comment by agilebean** [comment]
@andenixa thanks a lot - forgive me to ask again, how did you 
- enable histogram matching
- mask erosion
- smoothing
with the faceswap.py command? I didn't see these options, the closest was perceptual loss...

**Comment by HelpSeeker** [comment]
Those are options during the conversion, not the training. -mh activates histogram matching and with -e you can set the erosion kernel size.

**Comment by agilebean** [comment]
Hello @andenixa and @HelpSeeker,
thanks so much for your tips. 
The histogram matching and seamless options made a big difference.
As for mask erosion with the -e option, how do you specify 15%? 
`-e .15`or `-e 15` or else?

**Comment by andenixa** [comment]
Hello @agilebean,
I apologize for confusion there is now actual way at the moment to to supply parameters in percent. 
As of now you could only supply erosion kernel in pixels. What I meant is an approximate amount obtained by visual comparison of the results. I make in approximately about 15% of the overall mask size (20% in some difficult cases) by eye comparison.

**Comment by tjdwo13579** [comment]
So the moire pattern problem is still not solved right?
Cause I've witnessed this after 60k iterations with 32 batch

![image](https://user-images.githubusercontent.com/36047765/42411891-7e037b0c-823f-11e8-8913-0fe0e3ba77f2.png)

**Comment by andenixa** [comment]
@tjdwo13579 I don't want to overshadow Shaonlu's work in any way. But at FaceSwap none ever got good results with GAN2.1 (64x nor 128x). I rained it for a week with no conclusive results.

**Comment by tjdwo13579** [comment]
@andenixa Ah I see. Thanks for the reply.

I thought this was solved since there was a 3 month period.

So the result you showed on the Donald trump -> Cage video is using Shaonlu's repository?

**Comment by andenixa** [comment]
@tjdwo13579 no, it was DeepFaker (a regular autoencoder, no GAN involved).

**Comment by tjdwo13579** [comment]
@andenixa  Oops sorry I didn't read your post thoroughly.

**Comment by andenixa** [comment]
@tjdwo13579 I think the moire pattern is due to initial initialization with random pattern which propagates all layers and mask "thinks" this pattern is actual a part of the deal since Shaonlu mixes mask with the results and there is no separate (as far as I know) loss function for it. You could make different weights initializations as a temporary fix. I could play with it actually, but not sure if PR would ever get accepted.

**Comment by agilebean** [comment]
@andenixa sorry could your clarify your answer:

> Hello @agilebean,
> I apologize for confusion there is no actual way at the moment to to supply parameters in percent.
> As of now you could only supply erosion kernel in pixels. What I meant is an approximate amount > > obtained by visual comparison of the results. I make in approximately about 15% of the overall face > dimensions (20% in some difficult cases) by eye comparison.

So concretely, how do you specify the erosion kernel in pixels?
Do you specify a matrix, and if yes, in which format? If not, is is absolute amount of pixels?
Sorry to bother again, but there is nothing in the online help that specifies exactly how you specify each option.

**Comment by andenixa** [comment]
@agilebean 
I just use erosion and smooth via cmdline option and look at the merged results until borders take approx. 15% of mask size. The actual values vary depending on the image size and Model used.

**Comment by agilebean** [comment]
@andenixa 
thanks. now i got you. the value specified with option -e seems to be the number of pixels from the outer edge of the mask. now i understand that what you meant is the border width which is removed from the mask outer edge - and that's what is meant by erosion. 
Thanks again!
Just one wish for the future:
Can you provide these specifications in the --help ?
